Enhance and Extend HDF5 IO Functionality in IMASdd

mgyoo86 opened the pull request:
## Summary
This PR improves the HDF5 IO capabilities within IMASdd. 
Key functions such as `imas2hdf` and `h5merge` have been enhanced to support additional metadata, better file/group handling, and more flexible merging options. 
In addition, a new function `read_combined_h5` has been introduced for loading merged HDF5 files, and a new sample file (`omas_sample_with_attrs.h5`) has been added for testing. 
These improvements streamline both the writing and reading of IMAS data and enhance robustness when dealing with multiple files and directories.

## New Features

### `imas2hdf` Enhancements
- **File Attributes:**  
  - Automatically adds file-level attributes of the IMASdd version and current date/time (using the `Dates` package).
- **Group Attributes:**  
  - Each group now has attributes, such as `abstract_type` and `concrete_type`. This metadata can be used internally to automatically construct the appropriate IMAS object (eliminating the need for an external type parameter).
- **Additional Keywords:**  
  - E.g., a new keyword `desc` to provide descriptive information (e.g., shot number, description) that is stored as an attribute.

### `hdf2imas` Enhancements
- **New Dispatch Method:**  
  - A new dispatch of `hdf2imas` now accepts a `target_path` argument, enabling direct retrieval of a specific group or dataset from within the HDF5 file.
  - IDS object is automatically created inside the function, by using `concrete_type` attribute of the target group.

### `h5merge` Improvements
- **Include Base Directory Flag:**  
  - Introduces an `include_base_dir` flag to control whether the base directory is retained in the merged HDF5 structure.
- **Enhanced Merging Workflow:**  
  - Supports merging from both a single directory and multiple directories with consistent behavior.
  - Supports regex filtering (via `pattern` keyword) to merge specific files (e.g., only .h5 files)
- **Better Handling of Edge Cases**  

### New Function: `read_combined_h5`
 - Iteratively traverses a combined HDF5 file (starting at the root) and builds a dictionary mapping internal HDF5 paths to loaded data.
 - Supports regex filtering (via a `pattern` keyword) to selectively process only the desired datasets or groups.
- As it traverses, if it encounters an IDS object (determined via attribute checks) or a group whose name ends with `.h5`, it stops descending into that branch and loads that group as a whole; otherwise, it continues to traverse to extract all contained datasets.

## Test & Sample Updates

- **New Sample File:**  
  - `omas_sample_with_attrs.h5` has been added to the `sample/` directory. This file includes attributes, facilitating thorough testing of the new metadata features.
- **Test Suite Updates (in `test/runtests_io.jl`):**  
  - Added tests to verify:
    - File-level attributes (IMASdd version, date/time) and group attributes (abstract and concrete types).
    - The new `hdf2imas` dispatch method using a target path.
    - Enhanced `h5merge` functionality across single and multiple directories, with and without including the base directory.
    - Correct behavior of `read_combined_h5` with filtering (via regex) and retrieval of merged data.

mgyoo86 commented:
@orso82 FYI, I'll update this PR today to increase the coverage.

orso82 commented:
Of course. Thank you @mgyoo86 
Can you also cherry-pick this commit in your branch
https://github.com/ProjectTorreyPines/IMASdd.jl/commit/af42cc4246109da049723c03e60969b061312a8a

mgyoo86 commented:
@orso82 
## Updated Features and Improvements

- **Increased Test Coverage:**  
  - Added more tests to improve overall code coverage.

- **`cleanup` Keyword in `h5merge`:**  
  - Implemented a new `cleanup` keyword in the `h5merge` function.  
  - Introduced helper functions: `cleanup_files_and_directory` and `directory_contains_files`.  
  - The function first verifies that all target files have been successfully merged into `combined.h5`.  
  - Once verified, it cleans up the target files and, if the base folder is empty (i.e., contains no files even in its subdirectories), it deletes the base folder as well.

- **Tested on Omega:**  
  - Works well as expected.  
  - Note: The logical file size (as reported by `ls`) is about twice the actual disk usage (as reported by `du`), likely due to filesystem sparse allocation.

## `h5merge` Examples

```julia
# Merge files from a single directory:
h5merge("combined.h5", "dir_name")

# Merge files from multiple directories:
h5merge("combined.h5", ["dir1", "dir2"])

# Merge only files matching a specific pattern (e.g., only JSON files):
h5merge("combined.h5", ["dir1", "dir2"]; pattern=r"\.json$")

# Merge and cleanup: After merging, remove the target files and delete the base folder if empty.
h5merge("combined.h5", ["dir1", "dir2"]; cleanup=true)
```

orso82 commented:
Thank you for being so thorough @mgyoo86 !
